File: app/src/test/java/com/kickstarter/libs/utils/RewardUtilsTest.java
package com.kickstarter.libs.utils;

import android.content.Context;
import android.util.Pair;

import com.kickstarter.KSRobolectricTestCase;
import com.kickstarter.R;
import com.kickstarter.libs.KSString;
import com.kickstarter.libs.models.OptimizelyExperiment;
import com.kickstarter.mock.factories.LocationFactory;
import com.kickstarter.mock.factories.ProjectFactory;
import com.kickstarter.mock.factories.RewardFactory;
import com.kickstarter.models.Reward;

import org.joda.time.DateTime;
import org.joda.time.MutableDateTime;
import org.junit.Test;

import java.util.Date;

public final class RewardUtilsTest extends KSRobolectricTestCase {

  @Test
  public void testIsAvailable() {
    assertTrue(RewardUtils.isAvailable(ProjectFactory.project(), RewardFactory.reward()));
    assertFalse(RewardUtils.isAvailable(ProjectFactory.project(), RewardFactory.ended()));
    assertFalse(RewardUtils.isAvailable(ProjectFactory.project(), RewardFactory.limitReached()));
    assertFalse(RewardUtils.isAvailable(ProjectFactory.successfulProject(), RewardFactory.reward()));
    assertFalse(RewardUtils.isAvailable(ProjectFactory.successfulProject(), RewardFactory.ended()));
    assertFalse(RewardUtils.isAvailable(ProjectFactory.successfulProject(), RewardFactory.limitReached()));
  }

  @Test
  public void testDeadlineCountdownDetailWithDaysLeft() {
    final Context context = context();
    final KSString ksString = ksString();
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addDays(31);

    final Reward rewardWith30DaysRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownDetail(rewardWith30DaysRemaining, context, ksString), "days to go");
  }

  @Test
  public void testDeadlineCountdownDetailWithHoursLeft() {
    final Context context = context();
    final KSString ksString = ksString();
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addHours(3);

    final Reward rewardWith30DaysRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownDetail(rewardWith30DaysRemaining, context, ksString), "hours to go");
  }

  @Test
  public void testDeadlineCountdownDetailWithSecondsLeft() {
    final Context context = context();
    final KSString ksString = ksString();
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addSeconds(3);

    final Reward rewardWith30DaysRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownDetail(rewardWith30DaysRemaining, context, ksString), "secs to go");
  }

  @Test
  public void testDeadlineCountdownUnitWithDaysLeft() {
    final Context context = context();
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addDays(31);

    final Reward rewardWithDaysRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownUnit(rewardWithDaysRemaining, context), "days");
  }

  @Test
  public void testDeadlineCountdownUnitWithHoursLeft() {
    final Context context = context();
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addHours(3);

    final Reward rewardWithHoursRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownUnit(rewardWithHoursRemaining, context), "hours");
  }

  @Test
  public void testDeadlineCountdownUnitWithMinutesLeft() {
    final Context context = context();
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addMinutes(3);

    final Reward rewardWithMinutesRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownUnit(rewardWithMinutesRemaining, context), "mins");
  }

  @Test
  public void testDeadlineCountdownUnitWithSecondsLeft() {
    final Context context = context();
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addSeconds(30);

    final Reward rewardWithSecondsRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownUnit(rewardWithSecondsRemaining, context), "secs");
  }

  @Test
  public void testDeadlineCountdownValueWithMinutesLeft() {
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addSeconds(300);

    final Reward rewardWithMinutesRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownValue(rewardWithMinutesRemaining), 5);
  }

  @Test
  public void testDeadlineCountdownValueWithHoursLeft() {
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addSeconds(3600);

    final Reward rewardWithHoursRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownValue(rewardWithHoursRemaining), 60);
  }

  @Test
  public void testDeadlineCountdownValueWithDaysLeft() {
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addSeconds(86400);

    final Reward rewardWithDaysRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownValue(rewardWithDaysRemaining), 24);
  }


  @Test
  public void testDeadlineCountdownValueWithSecondsLeft() {
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addSeconds(30);

    final Reward rewardWithSecondsRemaining = RewardFactory.reward().toBuilder()
      .endsAt(currentDate.toDateTime())
      .build();
    assertEquals(RewardUtils.deadlineCountdownValue(rewardWithSecondsRemaining), 30);
  }

  @Test
  public void testHasBackers() {
    assertTrue(RewardUtils.hasBackers(RewardFactory.backers()));
    assertFalse(RewardUtils.hasBackers(RewardFactory.noBackers()));
  }

  @Test
  public void testIsLimited() {
    final Reward rewardWithRemaining = RewardFactory.reward().toBuilder()
      .remaining(5)
      .limit(10)
      .build();
    assertTrue(RewardUtils.isLimited(rewardWithRemaining));

    final Reward rewardWithNoneRemaining = RewardFactory.reward().toBuilder()
      .remaining(0)
      .limit(10)
      .build();
    assertFalse(RewardUtils.isLimited(rewardWithNoneRemaining));

    final Reward rewardWithNoLimitAndRemainingSet = RewardFactory.reward().toBuilder()
      .remaining(null)
      .limit(null)
      .build();
    assertFalse(RewardUtils.isLimited(rewardWithNoLimitAndRemainingSet));
  }

  @Test
  public void testIsItemized() {
    assertFalse(RewardUtils.isItemized(RewardFactory.reward()));
    assertTrue(RewardUtils.isItemized(RewardFactory.itemized()));
  }

  @Test
  public void testIsLimitReachedWhenLimitSetAndRemainingIsZero() {
    final Reward reward = RewardFactory.reward().toBuilder()
      .limit(100)
      .remaining(0)
      .build();
    assertTrue(RewardUtils.isLimitReached(reward));
  }

  @Test
  public void testIsLimitNotReachedWhenLimitSetButRemainingIsNull() {
    final Reward reward = RewardFactory.reward().toBuilder()
      .limit(100)
      .build();
    assertFalse(RewardUtils.isLimitReached(reward));
  }

  @Test
  public void testIsLimitReachedWhenRemainingIsGreaterThanZero() {
    final Reward reward = RewardFactory.reward().toBuilder()
      .limit(100)
      .remaining(50)
      .build();
    assertFalse(RewardUtils.isLimitReached(reward));
  }

  @Test
  public void testIsReward() {
    assertTrue(RewardUtils.isReward(RewardFactory.reward()));
    assertFalse(RewardUtils.isReward(RewardFactory.noReward()));
  }

  @Test
  public void testIsNoReward() {
    assertTrue(RewardUtils.isNoReward(RewardFactory.noReward()));
    assertFalse(RewardUtils.isNoReward(RewardFactory.reward()));
  }

  @Test
  public void testIsShippable() {
    final Reward rewardWithNullShipping = RewardFactory.reward()
      .toBuilder()
      .shippingType(null)
      .build();
    assertFalse(RewardUtils.isShippable(rewardWithNullShipping));

    final Reward rewardWithNoShipping = RewardFactory.reward();
    assertFalse(RewardUtils.isShippable(rewardWithNoShipping));

    final Reward rewardWithMultipleLocationShipping = RewardFactory.multipleLocationShipping();
    assertTrue(RewardUtils.isShippable(rewardWithMultipleLocationShipping));

    final Reward rewardWithSingleLocationShipping = RewardFactory.singleLocationShipping(LocationFactory.nigeria().displayableName());
    assertTrue(RewardUtils.isShippable(rewardWithSingleLocationShipping));

    final Reward rewardWithWorldWideShipping = RewardFactory.multipleLocationShipping();
    assertTrue(RewardUtils.isShippable(rewardWithWorldWideShipping));
  }

  @Test
  public void isTimeLimited() {
    assertFalse(RewardUtils.isTimeLimited(RewardFactory.reward()));
    assertTrue(RewardUtils.isTimeLimited(RewardFactory.endingSoon()));
  }

  @Test
  public void testIsExpired() {
    assertFalse(RewardUtils.isExpired(RewardFactory.reward()));
    final Reward rewardEnded2DaysAgo = RewardFactory.reward()
      .toBuilder()
      .endsAt(DateTime.now().minusDays(2))
      .build();
    assertTrue(RewardUtils.isExpired(rewardEnded2DaysAgo));
    final Reward rewardEndingIn2Days = RewardFactory.reward()
      .toBuilder()
      .endsAt(DateTime.now().plusDays(2))
      .build();
    assertFalse(RewardUtils.isExpired(rewardEndingIn2Days));
  }

  @Test
  public void testShippingSummary() {
    final Reward rewardWithNullShipping = RewardFactory.reward()
      .toBuilder()
      .shippingType(null)
      .build();
    assertNull(RewardUtils.shippingSummary(rewardWithNullShipping));

    final Reward rewardWithNoShipping = RewardFactory.reward();
    assertNull(RewardUtils.shippingSummary(rewardWithNoShipping));

    final Reward rewardWithMultipleLocationShipping = RewardFactory.multipleLocationShipping();
    assertEquals(Pair.create(R.string.Limited_shipping, null), RewardUtils.shippingSummary(rewardWithMultipleLocationShipping));

    final Reward rewardWithSingleLocationShipping = RewardFactory.singleLocationShipping(LocationFactory.nigeria().displayableName());
    assertEquals(Pair.create(R.string.location_name_only, "Nigeria"), RewardUtils.shippingSummary(rewardWithSingleLocationShipping));

    final Reward rewardWithWorldWideShipping = RewardFactory.rewardWithShipping();
    assertEquals(Pair.create(R.string.Ships_worldwide, null), RewardUtils.shippingSummary(rewardWithWorldWideShipping));
  }

  @Test
  public void minimumRewardAmountByVariant() {
    final Reward noReward = RewardFactory.noReward();
    assertEquals(1.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.CONTROL, noReward, 1));
    assertEquals(10.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_2, noReward, 1));
    assertEquals(20.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_3, noReward, 1));
    assertEquals(50.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_4, noReward, 1));
    assertEquals(10.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.CONTROL, noReward, 10));
    assertEquals(100.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_4, noReward, 100));
    assertEquals(100.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_3, noReward, 100));
    assertEquals(100.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_2, noReward, 100));
    assertEquals(100.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.CONTROL, noReward, 100));
    assertEquals(50.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_4, noReward, 5));
    assertEquals(10.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_2, noReward, 5));
    assertEquals(20.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_3, noReward, 5));
    assertEquals(5.0, RewardUtils.rewardAmountByVariant(OptimizelyExperiment.Variant.CONTROL, noReward, 5));
  }

  @Test
  public void testTimeInSecondsUntilDeadline() {
    final Date date = DateTime.now().toDate();
    final MutableDateTime currentDate = new MutableDateTime(date);
    currentDate.addSeconds(120);
    final Reward reward = RewardFactory.reward().toBuilder().endsAt(currentDate.toDateTime()).build();
    final long timeInSecondsUntilDeadline = RewardUtils.timeInSecondsUntilDeadline(reward);
    assertEquals(timeInSecondsUntilDeadline, 120);
  }
}

File: app/src/test/java/com/kickstarter/libs/utils/RewardUtilsTest.kt
package com.kickstarter.libs.utils

import android.content.Context
import android.util.Pair
import com.kickstarter.KSRobolectricTestCase
import com.kickstarter.R
import com.kickstarter.libs.KSString
import com.kickstarter.libs.models.OptimizelyExperiment
import com.kickstarter.libs.utils.RewardUtils.deadlineCountdownDetail
import com.kickstarter.libs.utils.RewardUtils.deadlineCountdownUnit
import com.kickstarter.libs.utils.RewardUtils.deadlineCountdownValue
import com.kickstarter.libs.utils.RewardUtils.hasBackers
import com.kickstarter.libs.utils.RewardUtils.hasStarted
import com.kickstarter.libs.utils.RewardUtils.isAvailable
import com.kickstarter.libs.utils.RewardUtils.isExpired
import com.kickstarter.libs.utils.RewardUtils.isItemized
import com.kickstarter.libs.utils.RewardUtils.isLimitReached
import com.kickstarter.libs.utils.RewardUtils.isLimited
import com.kickstarter.libs.utils.RewardUtils.isNoReward
import com.kickstarter.libs.utils.RewardUtils.isReward
import com.kickstarter.libs.utils.RewardUtils.isShippable
import com.kickstarter.libs.utils.RewardUtils.isTimeLimitedEnd
import com.kickstarter.libs.utils.RewardUtils.isTimeLimitedStart
import com.kickstarter.libs.utils.RewardUtils.isValidTimeRange
import com.kickstarter.libs.utils.RewardUtils.rewardAmountByVariant
import com.kickstarter.libs.utils.RewardUtils.shippingSummary
import com.kickstarter.libs.utils.RewardUtils.timeInSecondsUntilDeadline
import com.kickstarter.mock.factories.LocationFactory
import com.kickstarter.mock.factories.ProjectFactory
import com.kickstarter.mock.factories.RewardFactory
import com.kickstarter.models.Reward
import org.joda.time.DateTime
import org.joda.time.MutableDateTime
import org.junit.Before
import org.junit.Test
import java.util.*
import com.kickstarter.models.Project;

class RewardUtilsTest : KSRobolectricTestCase() {

    private lateinit var context: Context
    private lateinit var ksString: KSString
    private lateinit var date: Date
    private lateinit var currentDate: MutableDateTime
    private lateinit var reward: Reward
    private lateinit var noReward: Reward
    private lateinit var rewardEnded: Reward
    private lateinit var rewardLimitReached: Reward
    private lateinit var rewardMultipleShippingLocation: Reward
    private lateinit var rewardWorldWideShipping: Reward
    private lateinit var rewardSingleShippingLocation: Reward
    private lateinit var rewardWithAddons : Reward

    @Before
    fun setUpTests() {
        context = context()
        ksString = ksString()
        date = DateTime.now().toDate()
        currentDate = MutableDateTime(date)
        noReward = RewardFactory.noReward()
        reward = RewardFactory.reward()
        rewardEnded = RewardFactory.ended()
        rewardLimitReached = RewardFactory.limitReached()
        rewardMultipleShippingLocation = RewardFactory.multipleLocationShipping()
        rewardWorldWideShipping = RewardFactory.rewardWithShipping()
        rewardSingleShippingLocation = RewardFactory.singleLocationShipping(LocationFactory.nigeria().displayableName())
        rewardWithAddons = RewardFactory.rewardHasAddOns()
    }

    @Test
    fun testIsAvailable() {
        assertTrue(isAvailable(ProjectFactory.project(), reward))
        assertFalse(isAvailable(ProjectFactory.project(), rewardEnded))
        assertFalse(isAvailable(ProjectFactory.project(), rewardLimitReached))
        assertFalse(isAvailable(ProjectFactory.successfulProject(), reward))
        assertFalse(isAvailable(ProjectFactory.successfulProject(), rewardEnded))
        assertFalse(isAvailable(ProjectFactory.successfulProject(), rewardLimitReached))
    }

    @Test
    fun deadlineCountdownDetail_whenDaysLeft_returnsDayToGoString() {
        currentDate.addDays(31)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownDetail(reward, context, ksString), DAYS_TO_GO)
    }

    @Test
    fun deadlineCountdownDetail_whenHoursLeft_returnsHoursToGoString() {
        currentDate.addHours(3)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownDetail(reward, context, ksString), HOURS_TO_GO)
    }

    @Test
    fun deadlineCountdownDetail_whenSecondsLeft_returnsSecondsToGoString() {
        currentDate.addSeconds(3)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownDetail(reward, context, ksString), SECS_TO_GO)
    }

    @Test
    fun deadlineCountdownUnit_whenDaysLeft_returnsDays() {
        currentDate.addDays(31)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownUnit(reward, context), DAYS)
    }

    @Test
    fun deadlineCountdownUnit_whenHoursLeft_returnsHours() {
        currentDate.addHours(3)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownUnit(reward, context), HOURS)
    }

    @Test
    fun deadlineCountdownUnit_whenMinutesLeft_returnsMinutes() {
        currentDate.addMinutes(3)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownUnit(reward, context), MINS)
    }

    @Test
    fun deadlineCountdownUnit_whenSecondsLeft_returnsSecs() {
        currentDate.addSeconds(30)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownUnit(reward, context), SECS)
    }

    @Test
    fun deadlineCountdownValue_whenMinutesLeft_returnsNumberOfMinutes() {
        currentDate.addSeconds(300)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownValue(reward), 5)
    }

    @Test
    fun deadlineCountdownValue_whenHoursLeft_returnsNumberOfHours() {
        currentDate.addSeconds(3600)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownValue(reward), 60)
    }

    @Test
    fun deadlineCountdownValue_whenDaysLeft_returnsNumberOfDays() {
        currentDate.addSeconds(86400)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownValue(reward), 24)
    }

    @Test
    fun deadlineCountdownValue_whenSecondsLeft_returnsNumberOfSeconds() {
        currentDate.addSeconds(30)
        reward = RewardFactory.reward().toBuilder()
                .endsAt(currentDate.toDateTime())
                .build()
        assertEquals(deadlineCountdownValue(reward), 30)
    }

    @Test
    fun testHasBackers() {
        assertTrue(hasBackers(RewardFactory.backers()))
        assertFalse(hasBackers(RewardFactory.noBackers()))
    }

    @Test
    fun isLimited_whenRemainingGreaterThanZeroAndLimitNotNull_returnsTrue() {
        reward = RewardFactory.reward().toBuilder()
                .remaining(5)
                .limit(10)
                .build()
        assertTrue(isLimited(reward))
    }

    @Test
    fun isLimited_whenRemainingZeroAndLimitNotNull_returnFalse() {
        reward = RewardFactory.reward().toBuilder()
                .remaining(0)
                .limit(10)
                .build()
        assertFalse(isLimited(reward))
    }

    @Test
    fun isLimited_whenLimitAndRemainingNull_returnFalse() {
        reward = RewardFactory.reward().toBuilder()
                .remaining(null)
                .limit(null)
                .build()
        assertFalse(isLimited(reward))
    }

    @Test
    fun testIsItemized() {
        assertFalse(isItemized(reward))
        assertTrue(isItemized(RewardFactory.itemized()))
        assertTrue(isItemized(RewardFactory.itemizedAddOn()))
    }

    @Test
    fun isLimitReached_whenLimitSetAndRemainingIsZero_returnTrue() {
        reward = RewardFactory.reward().toBuilder()
                .limit(100)
                .remaining(0)
                .build()
        assertTrue(isLimitReached(reward))
    }

    @Test
    fun isLimitReached_whenLimitSetButRemainingIsNull_returnFalse() {
        reward = RewardFactory.reward().toBuilder()
                .limit(100)
                .build()
        assertFalse(isLimitReached(reward))
    }

    @Test
    fun isLimitReached_whenRemainingIsGreaterThanZero_returnFalse() {
        reward = RewardFactory.reward().toBuilder()
                .limit(100)
                .remaining(50)
                .build()
        assertFalse(isLimitReached(reward))
    }

    @Test
    fun testIsReward() {
        assertTrue(isReward(reward))
        assertFalse(isReward(noReward))
    }

    @Test
    fun testIsNoReward() {
        assertTrue(isNoReward(noReward))
        assertFalse(isNoReward(reward))
    }

    @Test
    fun testIsShippable() {
        val rewardWithNullShipping = RewardFactory.reward()
                .toBuilder()
                .shippingType(null)
                .build()
        assertFalse(isShippable(rewardWithNullShipping))
        assertFalse(isShippable(reward))
        assertTrue(isShippable(rewardMultipleShippingLocation))
        assertTrue(isShippable(rewardSingleShippingLocation))
        assertTrue(isShippable(rewardWorldWideShipping))
    }

    @Test
    fun isTimeLimited() {
        assertFalse(isTimeLimitedEnd(reward))
        assertTrue(isTimeLimitedEnd(RewardFactory.endingSoon()))
    }

    @Test
    fun isExpired_whenEndsAtPastDate_returnsTrue() {
        assertFalse(isExpired(reward))
        reward = RewardFactory.reward()
                .toBuilder()
                .endsAt(DateTime.now().minusDays(2))
                .build()
        assertTrue(isExpired(reward))
    }

    @Test
    fun isExpired_whenEndsAtFutureDate_returnsFalse() {
        reward = RewardFactory.reward()
                .toBuilder()
                .endsAt(DateTime.now().plusDays(2))
                .build()
        assertFalse(isExpired(reward))
    }

    @Test
    fun testShippingSummary() {
        val rewardWithNullShipping = RewardFactory.reward()
                .toBuilder()
                .shippingType(null)
                .build()
        val rewardWithNullLocation = RewardFactory.reward()
                .toBuilder()
                .shippingSingleLocation(null)
                .shippingType(Reward.SHIPPING_TYPE_SINGLE_LOCATION)
                .build()
        assertNull(shippingSummary(rewardWithNullShipping))
        assertNull(shippingSummary(reward))
        assertEquals(Pair.create(R.string.Limited_shipping, ""), shippingSummary(rewardWithNullLocation))
        assertEquals(Pair.create<Int, Any?>(R.string.Limited_shipping, null), shippingSummary(rewardMultipleShippingLocation))
        assertEquals(Pair.create(R.string.location_name_only, COUNTRY_NIGERIA), shippingSummary(rewardSingleShippingLocation))
        assertEquals(Pair.create<Int, Any?>(R.string.Ships_worldwide, null), shippingSummary(rewardWorldWideShipping))
    }

    @Test
    fun testRewardTimeLimitedStart_hasStarted() {
        val isLiveProject: Project = ProjectFactory.project()
        val rewardLimitedByStart = rewardWithAddons.toBuilder().startsAt(DateTime.now()).build()
        assertEquals(true, hasStarted(rewardLimitedByStart))
        assertEquals(true, isAvailable(isLiveProject, rewardLimitedByStart))
    }

    @Test
    fun testRewardNotTimeLimitedStart_hasStarted() {
        // - A reward not limited os starting time should be considered as a reward that has started
        assertEquals(false, isTimeLimitedStart(rewardWithAddons))
        assertEquals(true, hasStarted(rewardWithAddons))
    }

    @Test
    fun testRewardTimeLimitedStart_hasNotStarted() {
        val rewardLimitedByStart =
                rewardWithAddons
                        .toBuilder()
                        .startsAt(DateTime.now().plusDays(1))
                        .build()
        assertEquals(true, isTimeLimitedStart(rewardLimitedByStart))
        assertEquals(false, hasStarted(rewardLimitedByStart))
    }

    @Test
    fun testRewardTimeLimitedEnd_hasEnded() {
        val rewardExpired =
                rewardWithAddons
                        .toBuilder()
                        .endsAt(DateTime.now().minusDays(1))
                        .build()
        assertEquals(true, isExpired(rewardExpired))
        assertEquals(false, isValidTimeRange(rewardExpired))
    }

    @Test
    fun testValidTimeRage_limitedStart_hasNotStarted() {
        val rewardLimitedByStart =
                rewardWithAddons
                        .toBuilder()
                        .startsAt(DateTime.now().plusDays(1))
                        .build()
        assertEquals(false, isValidTimeRange(rewardLimitedByStart))
    }

    @Test
    fun testValidTimeRage_limitedStart_hasStarted() {
        val rewardLimitedByStart = rewardWithAddons.toBuilder().startsAt(DateTime.now()).build()
        assertEquals(true, isValidTimeRange(rewardLimitedByStart))
    }

    @Test
    fun testValidTimeRage_limitedEnd_hasNotEnded() {
        val rewardLimitedByEnd =
                rewardWithAddons
                        .toBuilder()
                        .endsAt(DateTime.now().plusDays(1))
                        .build()
        assertEquals(false, isExpired(rewardLimitedByEnd))
        assertEquals(true, isValidTimeRange(rewardLimitedByEnd))
    }

    @Test
    fun testValidTimeRange_limitedStartEnd_isValid() {
        val rewardLimitedBoth =
                rewardWithAddons
                        .toBuilder()
                        .startsAt(DateTime.now())
                        .endsAt(DateTime.now().plusDays(1))
                        .build()
        assertEquals(false, isExpired(rewardLimitedBoth))
        assertEquals(true, hasStarted(rewardLimitedBoth))
        assertEquals(true, isValidTimeRange(rewardLimitedBoth))
    }

    @Test
    fun testValidTimeRange_limitedStartEnd_isInvalid() {
        val rewardLimitedBoth =
                rewardWithAddons
                        .toBuilder()
                        .startsAt(DateTime.now().plusDays(1))
                        .endsAt(DateTime.now().plusDays(2))
                        .build()
        assertEquals(false, isExpired(rewardLimitedBoth))
        assertEquals(false, hasStarted(rewardLimitedBoth))
        assertEquals(false, isValidTimeRange(rewardLimitedBoth))
    }

    @Test
    fun isValidTimeRage_whenNotLimited_isValid() {
        assertEquals(false, isExpired(rewardWithAddons))
        assertEquals(true, hasStarted(rewardWithAddons))
        assertEquals(true, isValidTimeRange(rewardWithAddons))
    }

    @Test
    fun isValidTimeRage_whenStartNotLimited_returnsTrue() {
        assertEquals(true, isValidTimeRange(rewardWithAddons))
    }

    @Test
    fun minimumRewardAmountByVariant() {
        assertEquals(1.0, rewardAmountByVariant(OptimizelyExperiment.Variant.CONTROL, noReward, 1))
        assertEquals(10.0, rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_2, noReward, 1))
        assertEquals(20.0, rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_3, noReward, 1))
        assertEquals(50.0, rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_4, noReward, 1))
        assertEquals(10.0, rewardAmountByVariant(OptimizelyExperiment.Variant.CONTROL, noReward, 10))
        assertEquals(100.0, rewardAmountByVariant(OptimizelyExperiment.Variant.VARIANT_4, noReward, 100))
    }

    @Test
    fun testTimeInSecondsUntilDeadline() {
        currentDate.addSeconds(120)
        reward = RewardFactory.reward().toBuilder().endsAt(currentDate.toDateTime()).build()
        val timeInSecondsUntilDeadline = timeInSecondsUntilDeadline(reward)
        assertEquals(timeInSecondsUntilDeadline, 120)
    }

    companion object {
        private const val DAYS_TO_GO = "days to go"
        private const val HOURS_TO_GO = "hours to go"
        private const val SECS_TO_GO = "secs to go"
        private const val DAYS = "days"
        private const val HOURS = "hours"
        private const val MINS = "mins"
        private const val SECS = "secs"
        private const val COUNTRY_NIGERIA = "Nigeria"
    }
}

